### PR TITLE
Cognoma

### PR DESCRIPTION
**Pull request by chrsunwil**
I tried my best to add comments and change variable names to be more clear. The `Cognoma.py` file is still very hard to follow, but I'm not sure if I have a whole lot of ways to make that more clear. Please let me know if you have any ideas.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,7 @@ temp.py
 vae.py
 *.pyc
 __pycache__/
+*.tsv
+*.bz2
+*.csv
+pipeline.py
```
Comment (zietzm) [inline .gitignore]: Newline

File: Cognoma.py [hunk 2]
```diff
@@ -0,0 +1,662 @@
...
+        SNPpresent = SNPpresent[randIndexes]
+        SNPpresentExp = SNPpresentExp[randIndexes]
+
+        SNPabsent = SNPabsent[1:]
+        SNPabsentExp = SNPabsentExp[1:]
+        SNPabsent = SNPabsent[randIndexes]
+        SNPabsentExp = SNPabsentExp[randIndexes]
+
+        # Predict based on the mutations
+        (SNPpresentToExp, _) = modelHandle.rightToLeftModel.predict(
```
Comment (zietzm) [inline Cognoma.py:171]: Tuple parens not needed. Keep if you prefer that way

File: Cognoma.py [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+        plt.figure()
+        plt.bar(np.arange(LantentSpacePerc.shape[0]), LantentSpacePerc)
+        title = plt.title("Percentage difference between TP53 Wildtype and"
+                          " mutated for each latent space node, Mutation")
+        plt.xticks(np.arange(LantentSpacePerc.shape[0]), np.arange(
+            LantentSpacePerc.shape[0]))
+        plt.xlabel("Latent Space Nodes")
+        plt.ylabel("Latent Space Percentage Difference")
+        plt.savefig(os.path.join('Output', params.dataSetInfo.name,
+                                 'TP53MutLatentSpaceComparrison_{}.png'.
```
Comment (zietzm) [inline Cognoma.py:313]: If you're using Python 3.6+, could do 
```python
f'TP53MutLatentSpaceComparrison_{params.outputNum}.png'
```

File: Cognoma.py [hunk 4]
```diff
@@ -0,0 +1,662 @@
...
+        cluster = pd.DataFrame(X.transpose(), index=index, columns=y)
+
+        # Buils the color labels for cluster map
+        label_colors = np.repeat(sns.dark_palette("blue", 1, reverse=True),
+                                 num_examples, axis=0)
+
+        label_colors = np.append(label_colors,
+                                 np.repeat(sns.dark_palette("green", 1,
+                                                            reverse=True,),
+                                           num_examples, axis=0),
```
Comment (zietzm) [inline Cognoma.py:633]: I admire your dedication to proper indentation

File: DataSetInfoAbstractClass.py [hunk 5]
```diff
@@ -31,17 +31,20 @@ def load(self):
...
-                  left_generatedImgs, rightToLeftImgs, params, n=10):
+                  left_generatedImgs, rightToLeftImgs, params, n):
         """
         Visualizes all of the data passed to it. This does not need to return
         anything.
 
         Args:
-            randIndexes (array of ints): Random points to portray,
-                                         but same for each set of data
+            modelHandle (model): Holds all the components of the model
```
Comment (zietzm) [inline DataSetInfoAbstractClass.py:45]: Is `model` an actual type in this system? Would be nicer to have actual expected object type if not.

File: ICVL.py [hunk 6]
```diff
@@ -23,14 +23,13 @@
...
 from DataSetInfoAbstractClass import dataSetInfoAbstract
-import cPickle
 
 
 class dataInfo(dataSetInfoAbstract):
 
     def __init__(self):
         """
-        Defines the object's name, file locations and image size
+        Defines the object's name, file locations, image size and domain names
```
Comment (zietzm) [inline ICVL.py]: `Defines the object's name, file locations, image size, and domain names`

File: shared_vae_class.py [hunk 7]
```diff
@@ -67,250 +83,200 @@ def sampling(args):
...
+        # leftEncoderInputNorm = BatchNormalization()(leftEncoderInput)
         leftEncoderFirstLayer = Dense(
             self.params.firstLayerSizeLeft,
             activation='relu')(leftEncoderInput)
+        # leftEncoderFirstLayerNorm = BatchNormalization()
+        #                             (leftEncoderFirstLayer)
+        leftEncoderFirstLayerDropOut = Dropout(
+            self.params.dropout)(leftEncoderFirstLayer)
+
+        '''
```
Comment (zietzm) [inline shared_vae_class.py]: Why a string here?

**Review by zietzm — APPROVED** [review]
Nice. Two quick comments:

* Why Python 2.7?
* Hetmech recently ditched 79 character line limit because it was interfering with readability. Could switch to 120 if desired, because some indentation was a bit distracting.

File: DataSetInfoAbstractClass.py (hunk 5 above)
Comment (chrsunwil) [reply DataSetInfoAbstractClass.py:45]: model is the actual type, I could rename the type if that would make it easier to read.

File: Cognoma.py (hunk 3 above)
Comment (chrsunwil) [reply Cognoma.py:313]: I do need to switch to Python 3.6, there's no reason that I'm using 2.7 other than familiarity.

File: shared_vae_class.py (hunk 7 above)
Comment (chrsunwil) [reply shared_vae_class.py]: This is the second layer which I have chosen not to use, but I want to leave the code in so that it is easy to re-implement.

File: Cognoma.py [hunk 8]
```diff
@@ -0,0 +1,656 @@
...
+import copy
+
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+import pandas as pd
+import plotly as py
+import umap as up
+import plotly.graph_objs as go
+import sklearn as sk
```
Comment (gwaybio) [inline Cognoma.py]: it is best to not import the entire sklearn package, just import the modules you use (like you do in line 29)

File: Cognoma.py [hunk 9]
```diff
@@ -0,0 +1,656 @@
...
+                                          'Cognoma_Training_Large.pkl')
+        self.rightXDim = 45
+        self.rightYDim = 123
+        self.leftXDim = 100
+        self.leftYDim = 80
+        self.rightDomainName = "Mutation"
+        self.leftDomainName = "Expression"
+
+        # For Cognoma evaluation, which mutation to induce and compare
+        self.mutationID = 485
```
Comment (gwaybio) [inline Cognoma.py]: this mutation ID will not change ever? Should it belong as an `__init__` arg?

File: Cognoma.py [hunk 10]
```diff
@@ -0,0 +1,656 @@
...
+        Args: None
+
+        Returns: (Float array, Float array, Float array, Float array)
+                    The left training data, left testing data,
+                    right training data, right testing data
+        """
+        with open(self.training_file, "rb") as fp:
+            (x_temp, a_temp) = cPickle.load(fp, encoding='bytes')
+
+        # a_temp = preprocessing.scale(a_temp)
```
Comment (gwaybio) [inline Cognoma.py:72]: is the training file already scaled? worth adding scaling/normalization as method argument?

File: Cognoma.py [hunk 11]
```diff
@@ -0,0 +1,656 @@
...
+
+        # a_temp = preprocessing.normalize(a_temp, norm='l2')
+
+        x_temp, a_temp = sk.utils.shuffle(x_temp, a_temp)
+
+        length = x_temp.shape[0]
+
+        x_train = x_temp[:int(length * .9)]
```
Comment (gwaybio) [inline Cognoma.py]: lets use `train_test_split` for this: http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html

File: Cognoma.py (hunk 9 above)
Comment (gwaybio) [reply Cognoma.py]: Also, what does `485` represent? If entrez ID, TP53 is [`7157`](https://www.ncbi.nlm.nih.gov/gene/7157)

File: Cognoma.py [hunk 12]
```diff
@@ -0,0 +1,656 @@
...
+            params (model_parameters): Parameters of the model.
+            n (int): Defaults to 100
+
+        Returns: None
+        """
+
+        #######################################################################
+        # For calculating the t-test between the MSE of real and generated data
+
+        # Find all TP53 wildtype examples and their gene expressions
```
Comment (gwaybio) [inline Cognoma.py:137]: I am confused about what this is doing. Are you creating matrices for TP53 wildtype mutation and expression, or just getting the sample IDs? Can you use pandas to do this without looping through all samples?

File: Cognoma.py [hunk 13]
```diff
@@ -0,0 +1,656 @@
...
+            leftToRightCycle (array of floats): Left input
+                                               encoded and decoded as right,
+                                               then encoded and decoded as
+                                               left.
+            left_generatedImgs (array of floats): Random encoded points
+                                                    decoded as left.
+            rightToLeftImgs (array of floats): Right input encoded and decoded
+                                               as left.
+            params (model_parameters): Parameters of the model.
+            n (int): Defaults to 100
```
Comment (gwaybio) [inline Cognoma.py]: `n` needs to be more descriptive. perhaps `num_random` is better

File: Cognoma.py [hunk 14]
```diff
@@ -0,0 +1,656 @@
...
+        # Find the Mean Squared Error
+        SNPpresentToExpMSE = np.square(np.subtract(
+            SNPpresentExp, SNPpresentToExp)).mean(axis=0)
+
+        # Create arrays for induced samples
+        InducedMutation = copy.copy(SNPpresent)
+        InducedMutationExp = copy.copy(SNPpresentExp)
+
+        # Induce a TP53 mutation
+        InducedMutation[:, self.mutationID] = 1
```
Comment (gwaybio) [inline Cognoma.py:183]: this is inducing a mutation in all samples? or just a couple? Also, isn't the inducing happening to the `SNPpresent` data? Should it be happening to `SNPabsent`?

File: Cognoma.py [hunk 15]
```diff
@@ -0,0 +1,656 @@
...
+        length = x_temp.shape[0]
+
+        x_train = x_temp[:int(length * .9)]
+        x_test = x_temp[int(length * .9):]
+
+        a_train = a_temp[:int(length * .9)]
+        a_test = a_temp[int(length * .9):]
+        return (x_train, a_train, x_test, a_test)
+
+    def visualize(self, modelHandle,
```
Comment (gwaybio) [inline Cognoma.py]: This is a huge method! Can we brainstorm ways of splitting it up? How about we make smaller methods like `transform`, `induce_mutation`, `t_test_mse` and then a bunch of different visualization methods like `plot_latent_space_bar` and `plot_umap_reconstruction`, etc.. All of these functions can be split out into something more descriptive and will be easier to test.

File: Cognoma.py [hunk 16]
```diff
@@ -0,0 +1,656 @@
...
+
+        # Induce a TP53 mutation
+        InducedMutation[:, self.mutationID] = 1
+
+        # Predict based on this mutation
+        (InducedMutationToExp, _) = modelHandle.rightToLeftModel.predict(
+                                                             InducedMutation)
+
+        # Find the Mean Squared Error
+        InducedMutationToExpMSE = np.square(np.subtract(
```
Comment (gwaybio) [inline Cognoma.py]: can we use the sklearn implementation? http://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_squared_error.html

File: Cognoma.py [hunk 17]
```diff
@@ -0,0 +1,656 @@
...
+            SNPpresentToExpMSE, InducedMutationToExpMSE)
+        t_stat = ttest_results.statistic
+        p_val = ttest_results.pvalue
+
+        # Print a table with the t test results
+        table_data = dict(values=[[t_stat], [p_val]])
+        table_labels = dict(values=['Stat', 'PVal'])
+        table = [go.Table(cells=table_data, header=table_labels)]
+
+        py.offline.plot(table,
```
Comment (gwaybio) [inline Cognoma.py]: instead of plotting the table, lets write it to a text file instead. This can be done easily with pandas.

```python
import os
import pandas as pd

t_test_df = pd.DataFrame([t_stat, p_val], index=['t_stat', 'p_val']).transpose()

file = os.path.join('results', 't_test_mse_results.tsv')
t_test_df.to_csv(file, sep='\t')
```

File: Cognoma.py [hunk 18]
```diff
@@ -0,0 +1,656 @@
...
+                        filename=os.path.join('Output',
+                                              params.dataSetInfo.name,
+                                              f't_test_{str(params.outputNum)}'
+                                              ))
+        #######################################################################
+        # For Scatter Plot of the differentially expressed genes between
+        # real TP53 wildtype vs. real TP53 mutated and
+        # real vs. synthetic TP53 data
+
+        # Induce wiltype TP53
```
Comment (gwaybio) [inline Cognoma.py]: typo `wildtype`

File: Cognoma.py [hunk 19]
```diff
@@ -0,0 +1,656 @@
...
+        InducedSNPpresent = copy.copy(SNPabsent)
```
Comment (gwaybio) [inline Cognoma.py]: SNPabsent should have zeroes already no?

File: Cognoma.py [hunk 20]
```diff
@@ -0,0 +1,656 @@
...
+        InducedSNPpresent[:, self.mutationID] = 0
+
+        (InducedSNPpresentToExp, _) = modelHandle.rightToLeftModel.predict(
+                                                            InducedSNPpresent)
+
+        # Create y points for the linear regressions
+        xPoints = np.repeat(0, n)
+        xPoints = np.append(xPoints, np.repeat(1, n))
```
Comment (gwaybio) [inline Cognoma.py]: why do we need to do this? Is this compared against a null model?

Comment (gwaybio) [reply Cognoma.py]: I'll be able to comment further once I understand whats going on

File: Cognoma.py [hunk 21]
```diff
@@ -0,0 +1,656 @@
...
+        plt.savefig(os.path.join('Output', params.dataSetInfo.name,
+                                 f'Scatter_{str(params.outputNum)}.png'),
+                    bbox_extra_artists=(lgd,),
+                    bbox_inches='tight')
+
+        #######################################################################
+        # For creating bar graphs of latent space expression between
+        # TP53 Wiltype and TP53 mutated
+
+        # Find the difference in the nodes between TP53 wildtyupe and mutated
```
Comment (gwaybio) [inline Cognoma.py]: typo `wildtype`

**Review by gwaybio — CHANGES_REQUESTED** [review]
Very sorry for the extreme delay @chrsunwil - I hope this has not delayed your progress too much.

File: Cognoma.py (hunk 10 above)
Comment (chrsunwil) [reply Cognoma.py:72]: The training file is already scaled.

File: Cognoma.py (hunk 9 above)
Comment (chrsunwil) [reply Cognoma.py]: I can change it to use the entrez ID. The ID would only change if the user specified in the spot, so it could be in `__init__`

File: Cognoma.py (hunk 12 above)
Comment (chrsunwil) [reply Cognoma.py:137]: I'm creating matrices for the TP53 wildtype mutation and expression. I don't think I could do this without looping through all the sample. I'll check about if that would be possible in pandas, but I need to check the value in all examples.

File: Cognoma.py (hunk 14 above)
Comment (chrsunwil) [reply Cognoma.py:183]: It's only inducing in all samples where the SNP = 0, which I thought, based on conversations with others in the lab, meant that the SNP was present.

File: Cognoma.py (hunk 15 above)
Comment (chrsunwil) [reply Cognoma.py]: I thought about trying to split it up, but they all use the same data, so you would have to pass a massive amount of arguments around to all of the pieces.

Comment (chrsunwil) [reply Cognoma.py]: You can probably do some amount of breaking up, but I just wasn't sure if it would make it harder to follow because most of it builds off of the previous piece.

File: Cognoma.py (hunk 19 above)
Comment (chrsunwil) [reply Cognoma.py]: I thought that SNPabsent should have 1, and SNPpresent should have 0.

**Comment by chrsunwil** [comment]
I've addressed many of your comments. I haven't broken up the large visualization function because I think ti really does build off of previous pieces most of the time, and this preserves the logical ordering. It is, granted, a great candidate for a Jupyter notebook. That would probably be perfect for this.

File: DataSetInfoAbstractClass.py (hunk 5 above)
Comment (gwaybio) [reply DataSetInfoAbstractClass.py:45]: yeah, I agree that `model` is a confusing object name

File: ICVL.py [hunk 22]
```diff
@@ -56,10 +58,18 @@ def load(self):
...
+            (x_temp, a_temp) = pickle.load(fp, encoding='latin1')
 
-        with open(self.testing_file, "rb") as fp:
-            (x_test, a_test) = cPickle.load(fp)
+        np.random.shuffle(x_temp)
+        np.random.shuffle(a_temp)
+
+        length = x_temp.shape[0]
+
+        x_train = x_temp[:int(length * .9)]
```
Comment (gwaybio) [inline ICVL.py:68]: what is `0.9`? also can we use a more descriptive variable name than `length`?

File: README.md [hunk 23]
```diff
@@ -84,11 +104,20 @@ These are abstract functions, so a specific implimentation must be provided for
...
+It contains a load function which loads the data from a pickle file as well as a visualize function which produces images of the regular MNIST digits and the inverse MNIST digits.
+
+### `Cognoma.py`
+
+This is a specific implimentation of the `DataSetInfoAbstract` abstract class for the Cognoma data.
+It contains a load function which loads the data from a pickle file as well as a visualize function which produces many images and graphs.
+
+## Output
+
+The model saves various graphs and output into the specific folder for your data inside the `Output` folder. The model parameters will be saved in an `.html` with the number of the run serving as the title. All output will have the number of the run in the title. Additionally, diagrams of the left and right encoders will be saved, as will a line graph showing the training and validation loss over training. All other output will be determined by the `visualisation` method of it's data specific implementation of `DataSetInfoAbstract`.
```
Comment (gwaybio) [inline README.md:123]: Split this up - new sentences new lines

**Review by gwaybio — APPROVED** [review]
Nice updates @chrsunwil - a couple comments but then merge at will!
